exasol_advanced_analytics_framework/udf_communication/socket_factory/abstract_socket_factory.py

```diff
@@ -0,0 +1,99 @@
+import abc
+from enum import Enum, auto
+from typing import Union, List, Set, Optional
+
+
+class Frame(abc.ABC):
+    @abc.abstractmethod
+    def to_bytes(self) -> bytes:
+        pass
+
+
+class PollerFlag(Enum):
+    POLLIN = auto()
+    POLLOUT = auto()
+
+
+class Socket(abc.ABC):
+    @abc.abstractmethod
+    def send(self, message: bytes):
```

Comment:
if you would require anything that's byte convertible (implements `__bytes__`) you could just pass frames or raw data.
e.g.:
```python
class MagicSocket(Socket):

    def send(self, data):
        self.sock.send(bytes(data))
    ...

frame = IdFrame(id=1, data='some text')
sock = MagicSocket()

sock.send(frame)
sock.send(b"binary data")
...

```

Reply:
I am using different API for Frame and binary, because the underlying API distinguish them as well.